src/crimson/os/seastore/seastore_types.h

```diff
@@ -1025,7 +1026,128 @@ inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+ *
+ * # Static Layout
+ * laddr_t uses a 128-bit integer to represent the address value internally.
+ * Besides the basic properties inherits from integers, such as strong ordering
+ * and arithmetic operations, some types of laddr_t also includes the properties
+ * derived from user data. The static layout design ensures these properties are
+ * deterministic and predictable, allowing more optimization based on laddr_t.
+ *
+ * # Overview
+ * The layout of laddr_t has three parts:
+ *
+ * [upgrade:1][object_info:79][object_content:48]
+ *
+ * Each pair of square brackets represents a property stored within the laddr_t
+ * interger. Each property contains a name and the count of bits it used.
+ *
+ * When the object info for different objects does not conflict, we obtain a
+ * useful property:
+ * Each rados object and its head/clone could have a unique laddr prefix within
+ * an OSD.
+ *
+ * Base on this property, we could:
+ * 1. Group different snapshots of on object to the same laddr prefix, which
+ *    could speed up cloning process.
+ * 2. Track frequently accessed objects using laddr_t without keeping the full
+ *    object name.
+ * 3. If the rados object data/meta of a head/clone is under the same prefix,
+ *    removing a head/clone will also be possible via range deletion.
+ *
+ * # Object Info
+ * The defination of this property is:
+ * [pool:12][shard:8][reverse_hash:32][local_object_id:27]
+ *
+ * The pool, shard and reverse_hash come from the information of object;
+ * The local_object_id is just a random number to handle the collision between
+ * objects. When a collision occurs, the local clone ids of conflicting objects
+ * should be different.
+ *
+ * For global meta logical extents, such as RootBlock, CollectionNode and
+ * SeastoreNodeExtent, these fields should be set to zero, and other objects
+ * data should never use this prefix.
+ *
+ * # Object Content
+ * TODO: This property might change at runtime.
+ *
+ * The defination of this property is:
+ * [is_metadata:1][object_data:47]
+ *
+ * The `is_metadata` field is fixed and always located at the most significant
+ * bit of the object content, so it can't be changed at runtime.
+ *
+ * ## is_metadata true
+ * The remaining bits represent the address of Omap*Node. Take a random value
+ * as address when allocating a new omap extent.
+ *
+ * TODO: It's possiable for OMapInnerNode store only the value of this field
+ * instead of a full 128-bits laddr_t to increase the fan-out of the omap node.
+ *
+ * NOTE: Taking snapshots for omap extents follows the same steps as allocating
+ * a fresh omap extents, which doesn't use `local_clone_id` described below
+ * for object data extents. 47 bits should be enough for metadata usage.
+ *
+ * ## is_metadata false
+ * The remaining bits represent the address of ObjectDataBlock. It can be
+ * further divied into the following layout, which could also change at runtime:
+ *
+ * [local_clone_id:x][data_offset:47-x]
```

Comment:
The link https://github.com/ceph/ceph/commit/5c5d7423839f81dfec7fcadcbb59c934e7cd3d31#r1741633894 is broken, which comments do you refer?

Reply:
> So, probably we can rearrange the layout to make `x` static, **for example**:
> `[upgrade:1] [pool:12] [shard:4] [rhash:16] [local-object-id:31] [local-clone-id:32] [blocks:32]`
> So this allows:

IIRC, one pool can have at most 2^16 PGs, so there might be a relatively high probability that different PGs' data colide with the same `rhash`, if we configure it to have 16 bits.

Meanwhile, I think we need to consider cases in which objects are very small; for instance, we have rgw buckets in our online clusters that stores mainly 8KB-level objects. So, taking an extreme case: if objects are all 4KB-sized, a PG's logical storage space of the above design is `2^31 * 4KB = 8TB`, to avoid large amount of laddr conflicts, we'd better not let the occupation of its logical address space exceed `800GB`; on the other hand, according to our experience, clients keep making inappropriate `pg_num` configurations from time to time that lead to a relatively small number of PG per OSD, in which case the above design can lead to large amount of laddr conflicts or disk space wasting.

For objects' metadata(omap/xattr), it looks to us that there's no need to consider `local-clone-id`. Omaps/xattrs are stored in B-trees, which don't need `local-clone-id` to support COW.

What do you think?

Reply:
> The link https://github.com/ceph/ceph/commit/5c5d7423839f81dfec7fcadcbb59c934e7cd3d31#r1741633894 is broken, which comments do you refer?

I've fixed the link: it's the thread https://github.com/ceph/ceph/pull/59476#discussion_r1741633894

Reply:
> meanwhile, according to our experience, clients keep making inappropriate `pg_num` configurations from time to time that lead to a relatively small number of PG per OSD, in which case the above design can lead to large amount of laddr conflicts or disk space wasting.

The above design reduces the `[rhash]` field from 32 bits to 16 bits, so the saved 16 bits can be used to extend the address space for rados objects. Why does this lead to worse laddr conflicts or disk space wasting under small inappropriate `pg_num`?

Reply:
>For objects' metadata(omap/xattr), it looks to us that there's no need to consider local-clone-id. Omaps/xattrs are stored in B-trees, which don't need local-clone-id to support COW.

Replied in https://github.com/ceph/ceph/pull/59476#discussion_r1743884796

Reply:
> > meanwhile, according to our experience, clients keep making inappropriate `pg_num` configurations from time to time that lead to a relatively small number of PG per OSD, in which case the above design can lead to large amount of laddr conflicts or disk space wasting.
> 
> The above design reduces the `[rhash]` field from 32 bits to 16 bits, so the saved 16 bits can be used to extend the address space for rados objects. Why does this lead to worse laddr conflicts or disk space wasting under small inappropriate `pg_num`?

I meant that the number of PG per OSD was relatively small, for example, 2 PGs per OSD. In this case, according to the above calculation, at most `800GB * 2` of the disk space should be used if we want to avoid large amounts of conflicts.

> > So, taking an extreme case: if objects are all 4KB-sized, a PG's logical storage space of the above design is 2^31 * 4KB = 8TB, to avoid large amount of laddr conflicts, we'd better not let the occupation of its logical address space exceed 800GB;
> 
> IIUC the above design `[rhash:16] [local-object-id:31]` allows at most 2^47 objects per pool. As a theoretical upperbound, 2^47 should be sufficient, right?

Yeah, almost. `2 ^ 47` 4KB objects need totally 512PB storage space. To avoid large amounts of conflicts in this case, we might need 5EB laddr space. So, maybe we can adjust the layout as follows: `[local-object-id:35] [local-clone-id:32] [blocks:28]`, with at largest 1TB-sized head/clone objects and `512PB * 16` laddr space. What do you think?

Reply:
After an offline discussion with @cyx1231st , we came to the consensus about the layout as: `[local-object-id:43] [local-clone-id:24] [blocks:28]`. The main difference is that we shrinked the `local-clone-id` to 24 bits as an object's clone are highly unlikely to exceed 1M.

Reply:
^ The current decision of the 128 bit laddr is:
`[upgrade:1] [pool:12] [shard:4] [reverse_hash:16] [local_object_id:43] [local_clone_id:24] [is_metadata:1] [blocks:27]`
The upperbounds are roughly:
* pools conflicts at: 2^12 per cluster (4K)
* shard conflicts at: 2^4 per pool (16)
* pg conflicts at: 2^16 per pool&OSD (60K)
* unique objects: 2^59 per pool (1E)
* unique objects: 2^43 per PG (8T)
* unique clones: 2^24 per object (16M)
* data/metadata blocks at most: 2^27 per head/clone (128M)

Considering that allocations of objects/clones/metadata-blocks relies on random selection, and let's be conservative that the successful ratio should be >90% (There is room to optimize the selection algorithm if necessary), roughly:
* Data of an object: 512GB
* Metadata of an object: 50GB
* Clones of an object: 1M
* Extremely, even if all the rados objects are the smallest (4KB)
  * a PG: PBs of small objects
  * a pool: ZBs of small objects

Reply:
Is the shard field meant to be the pg's actual shard value, or do you plan to do an indirection?  The max shard value with 4 bits would be 15, right?  8+3 is a common configuration and with improvements to the EC implementation in RADOS, it wouldn't surprise me if highly capacity focused pools went wider than 16 shards.  A single OSD may rarely end up with more than 1 shard of the same PG, however, during rebalancing of a pretty small cluster.  It's pretty unlikely that it would end up with more than 2, though.

Reply:
There is no indirection design for now, let's extend shard field to 5 bits, the max shard value would be 31, and adjust the bits of `local_object_id` to 42.

Reply:
Seems to me the laddr `shard` field only needs to be (almost) unique in the pool to minimize transaction conflicts. So 4 bits wide should be sufficient if it is unlikely to be more than 2 in the same store, in a foreseeable future.

Implementing indirection seems to be simple by maintaining a small local map from the actual shard value to the local laddr shard field, which can be introduced later when necessary.

What do you think?

Reply:
> maintaining a small local map from the actual shard value to the local laddr shard field

Does it imply the read amplification of searching the mapping between actual shard value to local shard value?

Seems we should introduce a new extent type to store these mappings.

Reply:
It seems to be a small map which can be rebuilt from the onode tree upon restart -- I think we might be able to avoid to store this mapping in the disk, right?

Reply:
Oh, I get it. The only problem is memory usage, I think it's unnecessary to keep all objects' shard map in memory, only build this map via onode scan when constructing the seastore transaction.

Reply:
Yeah, it can be rebuilt on premise as well upon onode lookup. I suppose the memory usage is negligible at the pool-shard level.

Reply:
One thing needs to be clarified: can we build this map based on onode tree? At first I falsely thought the different shards of object are siblings on the onode tree, but it seems not, they locate at the different sub tree of onode tree. We can not construct this map by linear searching the onode tree within a small range, right?

Reply:
To build the map on start, for example, seems to me it is possible in the following steps:
1. Skip the global onode extents and lookup the first "normal" onode entry.
2. Add the pool-shard mapping to the map from the entry.
3. Set the lookup onode key to [pool] [shard] [MAX].
4. Lookup this key, it should return the first onode entry of the next valid pool-shard.
5. Repeat steps 2-4 until the end of the onode tree.

Does it make sense?

Reply:
> To build the map on start, for example, seems to me it is possible in the following steps:
> 
> 1. Skip the global onode extents and lookup the first "normal" onode entry.
> 2. Add the pool-shard mapping to the map from the entry.
> 3. Set the lookup onode key to [pool] [shard] [MAX].
> 4. Lookup this key, it should return the first onode entry of the next valid pool-shard.
> 5. Repeat steps 2-4 until the end of the onode tree.
> 
> Does it make sense?

I think the key of this mapping should be `[pgid][shard]`, because different PGs may have different shards in the same SeaStore instance and the total number of shards in one SeaStore instance may exceeds 16.

On the other hand, in SeaStore, we don't the total number of a pool's PGs, so we can't map an onode to a collection in SeaStore without changing SeaStore's API interface, which means we can't maintain this mapping in SeaStore based on the current SeaStore API interface.

How about we enlarge the `shard` field to 5 bits for now, and implement this mapping in the future? I think the number of shards shouldn't exceed 31 in most cases. What do you think? @cyx1231st @athanatos @Matan-B

Reply:
> I think the key of this mapping should be [pgid][shard], because different PGs may have different shards in the same SeaStore instance and the total number of shards in one SeaStore instance may exceeds 16.

Ah, so this breaks my very first assumption from https://github.com/ceph/ceph/pull/59476#discussion_r1743156748, quote:
> 1. It is unlikely for one SeaStore to store multiple active EC-shards of a pool, otherwise they will lost together. Probably we can reduce shard bits to 4 or even remove it?

Edit: it is normal for one SeaStore to store **every** EC-shards of a pool, because each PG places its shards independently among the OSDs.

If the edit above is correct, and if we still want to save the `shard` bits, probably the laddr layout should be adjusted to `[upgrade:1] [pool:12] [rhash:16] [local-shard-id:4] ...` in this PR?

(`local-shard-id` should be fine to start with the raw 5-bit `shard`, though).

Please note, this implies that the order of rados objects in the onode tree will be different from the laddr layout, e.g.:
```
The onode tree order:
shard-1 pool-1 PG-1 objects
shard-1 pool-1 PG-2 objects
shard-1 pool-2 PG-a objects
shard-1 pool-2 PG-b objects
shard-2 pool-1 PG-1 objects
shard-2 pool-1 PG-2 objects
shard-2 pool-2 PG-a objects
shard-2 pool-2 PG-b objects

The laddr layout:
pool-1 PG-1 shard-1 objects
pool-1 PG-1 shard-2 objects
pool-1 PG-2 shard-1 objects
pool-1 PG-2 shard-2 objects
pool-2 PG-a shard-1 objects
pool-2 PG-a shard-2 objects
pool-2 PG-b shard-1 objects
pool-2 PG-b shard-2 objects
```
I think the requirement to minimize transaction conflicts is to make sure each pool-PG-shard won't share the same lba subtree with the other pool-PG-shards, so the internal inconsistent laddr layout seems fine (Or am I missing anything here?). We have already reordered the shard-pool fields in this PR.

Reply:
@cyx1231st I think the problem lies in that we have to maintain the mapping from `[pgid][real_shard_id]` to `[local_shard_id]` in SeaStore while we don't know which pg an object belongs to. This is because SeaStore doesn't know the pg num of a pool which means it can't map objects to pgs.

Reply:
From the discussions, the current decision is to keep the `shard` field simple -- no redirection and keep it consistent with the onode tree order.